File: presentation/src/main/java/io/shelfy/presentation/details/DetailsGalleryFragment.java
package io.shelfy.presentation.details;

import android.os.Bundle;
import android.view.View;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.viewpager.widget.ViewPager;

import java.util.concurrent.atomic.AtomicInteger;

import io.shelfy.domain.entity.Movie;
import io.shelfy.presentation.BaseFragment;
import io.shelfy.presentation.R;
import io.shelfy.presentation.movies.viewmodel.MoviesViewModel;
import io.shelfy.presentation.movies.viewmodel.MoviesViewModelImpl;

public class DetailsGalleryFragment extends BaseFragment {

    private static final String ARGS_MOVIE_POSITION = "ARGS_MOVIE_POSITION";
    private static final String ARGS_MOVIE_ID = "ARGS_MOVIE_ID";

    private MoviesViewModel viewModel;
    private ViewPager moviesPager;
    private DetailsFragmentAdapter detailsFragmentAdapter;

    public DetailsGalleryFragment() {
        super(R.layout.fragment_movies_gallery);
    }

    public static DetailsGalleryFragment newInstance(Movie movie) {
        DetailsGalleryFragment galleryFragment = new DetailsGalleryFragment();
        final Bundle arguments = new Bundle();
        arguments.putInt(ARGS_MOVIE_ID, movie.getId());
        galleryFragment.setArguments(arguments);
        return galleryFragment;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        moviesPager = view.findViewById(R.id.vp_pager);
        detailsFragmentAdapter = new DetailsFragmentAdapter(getChildFragmentManager());
        moviesPager.setAdapter(detailsFragmentAdapter);

        final AtomicInteger currentPosition = new AtomicInteger();
        final AtomicInteger movieId = new AtomicInteger();
        final Bundle arguments = savedInstanceState != null ? savedInstanceState : getArguments();
        if (arguments != null) {
            movieId.set(arguments.getInt(ARGS_MOVIE_ID));
            currentPosition.set(arguments.getInt(ARGS_MOVIE_POSITION));
        }

        viewModel = fragmentComponent.getPresentationModule().provideViewModel(MoviesViewModel.class);
        viewModel.getMovies().observe(getViewLifecycleOwner(), movies -> {
            detailsFragmentAdapter.setData(movies);
            int currentItem = -1;
            int currentMovieId = movieId.get();
            if (currentMovieId != -1) {
                int index = 0;
                for (Movie movie : movies) {
                    if (movie.getId() == currentMovieId) {
                        currentItem = index;
                        break;
                    }
                    index++;
                }
            } else {
                currentItem = currentPosition.get();
            }
            if (currentItem != -1) {
                moviesPager.setCurrentItem(currentItem, false);
            }
        });
        viewModel.getErrorMessages().observe(
                getViewLifecycleOwner(),
                message -> Toast.makeText(requireContext(), message, Toast.LENGTH_LONG).show());
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        outState.putInt(ARGS_MOVIE_POSITION, moviesPager.getCurrentItem());
        super.onSaveInstanceState(outState);
    }
}

File: app/src/main/java/io/shelfy/architecture/ui/DetailsGalleryFragment.kt
package io.shelfy.architecture.ui

import android.os.Bundle
import android.view.View
import android.widget.Toast
import androidx.core.os.bundleOf
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import com.vladk.architecture.Dependencies
import com.vladk.architecture.R
import com.vladk.architecture.presentation.MoviesViewModel
import com.vladk.architecture.presentation.MoviesViewModelFactory
import com.vladk.architecture.ui.adapters.DetailsFragmentAdapter
import kotlinx.android.synthetic.main.fragment_movies_gallery.*

class DetailsGalleryFragment : Fragment(R.layout.fragment_movies_gallery) {

    companion object {

        private const val ARGS_MOVIE_POSITION = "ARGS_MOVIE_POSITION"

        fun newInstance(position: Int): DetailsGalleryFragment {
            return DetailsGalleryFragment().apply {
                arguments = bundleOf(ARGS_MOVIE_POSITION to position)
            }
        }
    }

    private val viewModelFactory: MoviesViewModelFactory = Dependencies.moviesViewModelFactory

    private lateinit var viewModel: MoviesViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        viewModel = ViewModelProvider(this, viewModelFactory)
            .get(MoviesViewModel::class.java)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        val position = savedInstanceState
            ?.getInt(ARGS_MOVIE_POSITION) ?: arguments?.getInt(ARGS_MOVIE_POSITION) ?: 0

        viewModel.moviesLiveData.observe(viewLifecycleOwner, Observer { movies ->
            vp_pager.run {
                adapter = DetailsFragmentAdapter(childFragmentManager, movies)
                currentItem = position
            }
        })
        viewModel.errorLiveData.observe(viewLifecycleOwner, Observer { message ->
            Toast.makeText(requireContext(), message, Toast.LENGTH_LONG).show()
        })
    }

    override fun onSaveInstanceState(outState: Bundle) {
        outState.putInt(ARGS_MOVIE_POSITION, vp_pager.currentItem)
        super.onSaveInstanceState(outState)
    }
}
